src/implementations/junoswap/staking.rs

```diff
@@ -0,0 +1,172 @@
+
+#[cw_serde]
+pub struct JunoswapStaking {
+    pub addr: Addr,
+    pub lp_token_addr: Addr,
+}
+
+impl Stake for JunoswapStaking {
+    fn stake(&self, deps: Deps, _env: &Env, amount: Uint128) -> Result<Response, CwDexError> {
+        let cfg = deps.querier.query::<Config>(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: self.addr.to_string(),
+            msg: to_binary(&Cw20StakeQueryMsg::GetConfig {})?,
+        }))?;
+
+        let stake_msg = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: cfg.token_address.to_string(),
+            funds: vec![],
+            msg: to_binary(&Cw20ExecuteMsg::Send {
+                contract: self.addr.to_string(),
+                amount,
+                msg: to_binary(&Cw20StakeReceiveMsg::Stake {})?,
+            })?,
+        });
+
+        let event = Event::new("cw-dex/staking/stake")
+            .add_attribute("type", "junoswap")
+            .add_attribute("amount", amount.to_string());
+
+        Ok(Response::new().add_message(stake_msg).add_event(event))
+    }
+}
+
+impl Unstake for JunoswapStaking {
+    fn unstake(&self, deps: Deps, env: &Env, amount: Uint128) -> Result<Response, CwDexError> {
+        // Verify that the vault does not have an unbonding period. Our design assumes
+        // that the vault does not have an unbonding period when unstake can be called.
+        let cfg = deps.querier.query::<Config>(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: self.addr.to_string(),
+            msg: to_binary(&Cw20StakeQueryMsg::GetConfig {})?,
+        }))?;
+
+        if cfg.unstaking_duration.is_some() {
+            return Err(CwDexError::UnstakingDurationNotSupported {});
+        }
+
+        // Locked and non locked staking uses the same unstake message on Junoswap.
+        self.unlock(deps, env, amount)
+    }
+}
+
+impl Rewards for JunoswapStaking {
+    fn claim_rewards(&self, deps: Deps, _env: &Env) -> Result<Response, CwDexError> {
+        let claim_messages = deps
+            .querier
+            .query::<GetHooksResponse>(&QueryRequest::Wasm(WasmQuery::Smart {
+                contract_addr: self.addr.to_string(),
+                msg: to_binary(&Cw20StakeQueryMsg::GetHooks {})?,
+            }))?
+            .hooks
+            .iter()
+            .map(|addr| {
+                // Call as SubMsg since we don't know if the contracts in the hooks are
+                // always going to be reward contracts. If not the messages for that
+                // contract should fail without failing the transaction.
+                Ok(SubMsg {
+                    id: 0,
+                    msg: CosmosMsg::Wasm(WasmMsg::Execute {
+                        contract_addr: addr.to_string(),
+                        funds: vec![],
+                        msg: to_binary(&StakeCw20ExternalRewardsExecuteMsg::Claim {})?,
+                    }),
+                    gas_limit: None,
+                    reply_on: ReplyOn::Never,
```

Comment:
Should be `ReplyOn::Always`?

Reply:
Even better is `ReplyOn::Error`, which would use less gas since we don't want to handle the reply anyway. https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md?plain=1#L248